pkg/client/client.go

```diff
@@ -0,0 +1,330 @@
+	// struct pointer so that obj can be updated with the content returned by the Server.
+	Update(ctx context.Context, obj runtime.Object) error
+
+	// Delete deletes the given obj from Kubernetes cluster.
+	Delete(ctx context.Context, obj runtime.Object) error
+
+	// List retrieves list of objects for a given namespace and list options. On a
+	// successful call, Items field in the list will be populated with the
+	// result returned from the server.
+	List(ctx context.Context, namespace string, opts metav1.ListOptions, list runtime.Object) error
+}
+
+// NewForConfig returns a Kubernetes client implementation given a
+// configuration.
+func NewForConfig(config *rest.Config) (Interface, error) {
+	if config == nil {
+		return nil, fmt.Errorf("missing config")
+	}
+	config = rest.CopyConfig(config)
+	cs, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	cachedDiscoveryClient := cached.NewMemCacheClient(cs.Discovery())
+	restMapper := restmapper.NewDeferredDiscoveryRESTMapper(cachedDiscoveryClient)
+	restMapper.Reset()
+
+	dc, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &client{
+		conf:       config,
+		dc:         dc,
+		scheme:     scheme.Scheme,
+		restMapper: restMapper,
+	}
+	// c.scheme.AddToScheme(scheme.Scheme)
+	c.runBackgroundCacheReset(1 * time.Minute)
+	return c, nil
+}
+
+// Client is a concrete implementation of Kubernetes Çlient interface.
+type client struct {
+	conf       *rest.Config
+	dc         dynamic.Interface
+	scheme     *runtime.Scheme
+	restMapper *restmapper.DeferredDiscoveryRESTMapper
+}
+
+// runBackgroundCacheReset - Starts the rest mapper cache reseting
+// at a duration given.
+func (c *client) runBackgroundCacheReset(duration time.Duration) {
+	ticker := time.NewTicker(duration)
+	go func() {
+		for range ticker.C {
+			c.restMapper.Reset()
+		}
+	}()
+}
+
+func (c *client) Create(ctx context.Context, obj runtime.Object) error {
+	ns, _, err := namespaceAndName(obj)
+	if err != nil {
+		return err
+	}
+	gvks, _, err := c.scheme.ObjectKinds(obj)
+	if err != nil {
+		return err
+	}
+	if len(gvks) == 0 {
+		return fmt.Errorf("object is not registered")
+	}
+	gvk := gvks[0]
+
+	resource, err := apiResource(gvk, c.restMapper)
+	if err != nil {
+		return fmt.Errorf("failed to get resource type: %v", err)
+	}
+
+	result, err := c.dc.Resource(resource).Namespace(ns).Create(unstructuredFromRuntimeObject(obj))
+	if err != nil {
+		return err
+	}
+	err = unstructuredToObject(result, obj)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *client) Get(ctx context.Context, key ObjectKey, obj runtime.Object) error {
+	// create dynamic client on the fly (worry about efficiency later)
+
+	// TODO(droot): figure out why GetObjectKind() is not working for an empty
+	// object ? Shouldn't it use scheme internally. If not then, there isn't
+	// much benefit in taking the runtime.Object
```

Comment:
I should have pointed this out earlier. We had to enforce specifying the TypeMeta for the object in our implementation:
```Go
pod := &v1.Pod{
	TypeMeta: metav1.TypeMeta{
		Kind:       "Pod",
		APIVersion: "v1",
	},
}
podKey := client.ObjectKey{Namespace: "ns-1", Name: "pod-1"}
err = kc.Get(ctx, podKey, pod)
if err != nil {
	// handle err
}
```
That makes it more cumbersome to use though. Using the scheme to get the GVK is fine if it lets us pass in an empty object.

However we still need to pass in the runtime.Object to unmarshal the unstructured object into it. I don't think we can do away with that.

Reply:
> That makes it more cumbersome to use though. Using the scheme to get the GVK is fine if it lets us pass in an empty object.

Yes, thats the idea. Using scheme, we are able to work with empty object and caller code looks very simple. Yes, we can't do away with the runtime.obj and the good thing all generated types support runtime.Obj interface, so we are good.